[WIP] First sample app proposal

brianegan opened the pull request:
We've taken the feedback from the community and we're starting to assemble it into a sample app. Please note, this is a work in progress!

**Features**

  - [x] List/Detail App with 2 Features: Items and Settings
  - [x] Localize the title of the app
  - [x] Analysis options using pedantic
  - [x] MVC with ChangeNotifier
  - [x] Only Flutter/Dart Team dependencies
  - [x] Objects passed down via constructors, no third-party library
  - [x] Feature-first folder structure
  - [x] 1 Image asset in 1x, 2x, and 3x 
  - [x] Dummy Tests that demonstrate Widget, Unit, and Integration Tests
  - [x] Logging & Error Reporting Implementation
  - [x] Responsive Design

**Refinement**

Tasks we want to address once we've agreed to the overall features & structure. 

  - [x] Add high-quality code comments throughout the app that follow Flutter official documentation style
  - [x] Clean up pubspec.yaml
  - [x] Use Widgets instead of `_buildMobileContent` style methods
  - [x] Rename Widgets for clarity 
  - [x] Finalize Enabled & Disabled Lint Rules & fix any related errors.
  - [x] Enable null safety and fix related errors? (Reviewed notes: No).

brianegan committed:
WIP First proposal

filiph commented on proposals/proposal_1/analysis_options.yaml:
```diff
@@ -0,0 +1,13 @@
+# Static analysis allows you to find problems before executing a single line of code.
+# You may customize the Analysis Options by following this guide:
```
```suggestion
# You may customize by following this guide:
```

filiph replied:
(Not sure about this.)

filiph commented on proposals/proposal_1/analysis_options.yaml:
```diff
@@ -0,0 +1,13 @@
+
+linter:
+  rules:
+    prefer_const_constructors: true
```
Maybe add one that says `: false` ?

brianegan replied:
I'm thinking of the following rules:

```yaml
linter:
  rules:
    prefer_void_to_null: true
    always_specify_types: false
```

`prefer_const_constructor` is nice, but it is very heavy handed and can produce errors that cover a lot of code. For an advanced dev, it's easy to see where to put the `const` constructor. For less experience devs, I think it will create friction.

Also, I know you wanted to disable the single-quoted rule, but it seemed odd to use pedantic and then override the rules it defines? This is just a suggestion and I don't have a strong opinion, so if you have two rules here you feel strongly about and would like to use please let me know :)

filiph replied:
That SGTM

filiph commented on proposals/proposal_1/lib/main.dart:
```diff
@@ -0,0 +1,44 @@
+    final settingsController = SettingsController(SettingsService());
+
+    // Load the user's settings before displaying the App.
+    await settingsController.loadSettings();
```
[nit] Do intermediate-to-large apps do this? Don't they try to show something first, and load settings later?

brianegan replied:
As with everything, it depends :)

If you're storing the settings on disk, I'd say you almost always want to load those setting before running the app. Otherwise you may have a "Flash of Unthemed Content" -- meaning you may show the app in the default theme, then load the user settings very quickly from disk while showing a spinner, then change everything to a different theme. That can be a pretty quick and jarring transition, especially if you're going from a default light theme to a dark theme, or vice-versa.

Instead, it's generally better from a user experience standpoint to let the Splash screen show for an additional 5-10ms to load the settings from disk, then show the app with the correct theme from the get-go. This avoids any big theme changes.

However, if you're loading settings from the network, this technique is not ideal.

filiph replied:
You're right. Let's keep with this. You can always update settings from the network later (and have FOUC then) but generally, this makes sense.

filiph commented on proposals/proposal_1/lib/main.dart:
```diff
@@ -0,0 +1,44 @@
+import 'package:proposal_1/src/settings/settings_service.dart';
+
+void main() {
+  // Set up cross-cutting Services, such as Logging, Analytics, or Feature Flags
```
I like this.

filiph commented on proposals/proposal_1/lib/main.dart:
```diff
@@ -0,0 +1,44 @@
+    await settingsController.loadSettings();
+
+    runApp(MyApp(
+      itemController: ItemController(ItemService(), loggingService),
```
I'd extract `ItemController` above, so it sits next to the other controllers (`final itemController = ItemController(...)`). Also, shouldn't it be `DummyController`?

filiph commented on proposals/proposal_1/lib/main.dart:
```diff
@@ -0,0 +1,44 @@
+
+    runApp(MyApp(
+      itemController: ItemController(ItemService(), loggingService),
+      settingsController: settingsController,
```
Looking at this, I'm wondering if there's another option we didn't think about:

1) Set up everything like you have here: settings, logging, runZoned. These are needed in basically _every_ intermediate-to-large app.
2) Do _not_ set up any `DummyController`, `ItemController` or anything similar. That's code that will need to be deleted. Instead, use `SettingsController` (and possibly `LoggingController`) to show how to structure things.

I'm not done reading, just throwing this out there.

brianegan replied:
Removed!

filiph commented on proposals/proposal_1/lib/src/app.dart:
```diff
@@ -0,0 +1,57 @@
+  @override
+  Widget build(BuildContext context) {
+    // Listen to the SettingsController for changes to the App Theme. Rebuild
+    // the App whenever the theme changes
```
[nit] I'm used to periods (`.`) at end of comments. Is omitting periods a style or is it just a typo here?

brianegan replied:
Yah sorry -- these comments are really rough overall! Wanted to give a general idea without documenting everything since I knew we'd change up a lot of this code.

I'll make sure to re-read and follow the Flutter guidelines for code comments for the final product.

brianegan replied:
I've tried to add high quality comments throughout the codebase. I think I've fixed formatting and adhered to proper sentence structure, but for some reason my grammar in comments is maddeningly inconsistent!

I appreciate you reading through the comments with an eye for detail :)

filiph commented on proposals/proposal_1/lib/src/app.dart:
```diff
@@ -0,0 +1,57 @@
+              builder: (context) {
+                if (routeSettings.name == '/settings') {
+                  return SettingsView(controller: settingsController);
+                } else if (routeSettings.name.contains('item')) {
```
```suggestion
                } else if (routeSettings.name.startsWith('/item')) {
```

Maybe?

filiph commented on proposals/proposal_1/lib/src/app.dart:
```diff
@@ -0,0 +1,57 @@
+                  return SettingsView(controller: settingsController);
+                } else if (routeSettings.name.contains('item')) {
+                  return ItemDetailsPage(
+                    id: int.parse(routeSettings.name.split('/').last),
```
On one hand, I get this. We're trying to make something work. On the other hand, this is so far from what a real app would be doing (you don't want to crash when someone navigates to `/item/23/`, for example) that I'm once again thinking if we want to just handwave this.

```suggestion
                    // TODO: parse routeSettings.name to actually extract the number.
                    id: 42,
```

Ugh, that's also bad. Just food for though. I'll also be thinking about it.

filiph commented on proposals/proposal_1/lib/src/items/item.dart:
```diff
@@ -0,0 +1,5 @@
+class Item {
```
DummyItem? Not sure about this. Item is also good.

filiph commented on proposals/proposal_1/lib/src/items/item_controller.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:proposal_1/src/items/item_service.dart';
+import 'package:proposal_1/src/logging/logging_service.dart';
+
+// A class that many Widgets can interact with to read a List of Item objects,
+// update a List of Item objects, or listen to changes.
+//
+// Controllers glue Data Services to Flutter Widgets.
```
Love this.

filiph replied:
I would add something like:

```dart
/// To implement a controller, you can use Flutter's [ChangeNotifier] class, like here, or you can
/// use its equivalent in your favorite state management package: <list some examples here>
```

Also, please use doc comments (`///`), not regular ones (`//`).

brianegan replied:
Do we want to use Doc Comments here? I wasn't sure if this was the type of content that should end up in documentation generated by `dartdoc`, or if it was just a comment for the reader.

brianegan replied:
From meeting: using doc comments is a good idea!

However, I thought about adding the additional comment to SettingsController, but I'm actually wondering if we should leave it out. From the opening paragraphs of the requirements document: "that leap lands them among a myriad of community solutions and opinions, most of which a novice user is not yet ready to review critically."

I kinda think that's right after interviewing some folks -- many folks aren't quite ready to critically review them. I'm wondering if it makes more sense to point them to the Simple State Management tutorial instead so they have more context on how the Controller is being used?

filiph commented on proposals/proposal_1/lib/src/items/item_controller.dart:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  Future<void> addItem(Item item) async {
+    _items = [..._items, item];
```
I'm curious why not just use `_items.add`. It would remove a syntax feature from the cognitive load.

filiph commented on proposals/proposal_1/lib/src/items/item_controller.dart:
```diff
@@ -0,0 +1,69 @@
+    try {
+      _items = await itemService.items();
+      loggingService.debug('Successfully loaded Items');
+    } catch (error) {
```
I wonder if this looks like we're teaching/showing an anti-pattern. At the very least:

```suggestion
    } catch (exception) {
```

But, ideally, we know what exception(s) to catch.

brianegan replied:
Can you describe which anti-pattern in a bit more detail? I had an idea I was trying to demonstrate here, but it sounds like it didn't work out :) 

However, I think I'm thinking of a different anti-pattern than you, haha, so I'd be curious what you specifically meant here.

brianegan replied:
Nevermind, deleted :)

filiph commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+        if (widget.controller.loading) {
+          return const Center(child: CircularProgressIndicator());
+        } else if (widget.controller.error) {
+          return const Center(child: Text('Oh no! Could not load the users'));
```
```suggestion
          return const Center(child: Text('Oh no! Could not load items'));
```

filiph commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+            ),
+            body: constraints.maxWidth < 600
+                ? _buildMobileContent(context)
+                : _buildTabletContent(context),
```
```suggestion
                ? _buildNarrowContent(context)
                : _buildWideContent(context),
```

Maybe? We're not only supporting mobile devices.

filiph commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+          return const Center(child: Text('Oh no! Could not load the users'));
+        }
+
+        return LayoutBuilder(builder: (context, constraints) {
```
Is there a reason to have the `LayoutBuilder` up top like this? Shouldn't it be around the `body:`? Or is this an extensibility thing?

brianegan replied:
You got it! After trying a number of different responsive designs, placing the LayoutBuilder over the whole Scaffold felt like the easiest option for folks to modify and extend. For example, the current design places an AppBar over the List and Item Details. However, many apps only display an AppBar above the item details instead. Where to put the FloatingActionButton can become another fun game :)

Overall, I really wonder if we should include responsive breakpoints & the corresponding code. The more I played around with this,  the more I began to think that developers will end up deleting most of this and rewriting it depending on the design of their app.

  > Can we build the functionality in a way that 90% of developers will want to keep it and/or add to it, instead of ripping it out?

Where the LayoutBuilder needs to appear in the Widget hierarchy, which breakpoints they want to handle, etc all heavily depend to the unique design of their app and the goals for the organization. What do you think?

filiph replied:
That's fair. I'm glad we have the data now. Let's discuss.

filiph commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  Widget _buildMobileContent(BuildContext context) {
+    return _ListView(
```
The naming here is hard. It could be confused with some secret variant of `ListView`. Nit.

filiph commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+            onItemTapped: _onItemTapped,
+          ),
+        ),
+        Container(
```
could be a VerticalDivider?

filiph commented on proposals/proposal_1/lib/src/items/item_service.dart:
```diff
@@ -0,0 +1,15 @@
+// A Service that Manages Item Objects
+class ItemService {
+  Future<List<Item>> items() {
+    return Future.delayed(
```
Add a comment saying something like:

```dart
// Adding a fake delay so that the app behaves more like a real one.
```

filiph commented on proposals/proposal_1/lib/src/logging/logging_service.dart:
```diff
@@ -0,0 +1,25 @@
+///   - Debug version: Prints information to the console as you work. You
+///     should only modify this file if you want to change how to print messages
+///     during development.
+///   - Profile version: Does nothing. Printing or sending logs to a service may
+///     cause unwanted jank during a performance profile. You should only change
+///     this file if you want to modify how logs are handled during performance
+///     profiles.
```
Interesting. I would worry a bit that apps that do this will discount the actual cost of logging: their profiling will never uncover logging-related performance issues. But I understand the desire to remove the fluctuation, especially for microbenchmarks. 🤷

brianegan replied:
Another "it depends" haha.

I usually disable printing in profile mode, since it slows down and may add noise to the benchmark results. 

I also disable reporting to a service in profile mode, since profiles may run lots and lots of times on CI servers and you do not want to send that data to Crashlytics or Sentry.

filiph commented on proposals/proposal_1/lib/src/logging/logging_service.dart:
```diff
@@ -0,0 +1,25 @@
+/// A Service that records application messages or exceptions.
```
FWIW, I still think setting up `package:logging` is fine for the intermediate app, and it captures logs from (well setup) packages/plugins as well. Plus it has things like `log.fine(() => 'some expensive $string interpolation)`.

brianegan replied:
I'm happy to explore that alternative, but can you please describe your use cases in a bit more detail? Or perhaps demonstrate with code if that's easier than an explanation?

From my experience, you have a few use cases for logging:

  - Print logs in Debug
  - Report information to a service in release mode. 
  - Disable printing in profile mode, since it adds noise to the benchmark results. Disable reporting to a service in profile mode, since profiles may run many many times on CI servers and you do not want to send junk data to your logging service.

While package:logging  gives you something send logs to, it does not provide a structure to solve these particular use cases. Rather, it's a bit like a Logging Event Bus system. To perform the tasks I mentioned, you'd still need to do something like this:

```dart
Logger.root.onRecord.listen((record) {
 // perform all of the logic here to send the record to the correct backend depending on the environment.
  if (kDebugMode) {
    print('${record.level.name}: ${record.time}: ${record.message}');
  } else if (kReleaseMode) {
    switch (record.level) {
      case Level.Severe:
        // send it to sentry
    }
  }
});

// or Still use the service depending on the environment...
final loggingService = kDebugMode ? DebugLoggingService()...;

Logger.root.onRecord.listen((record) {
  switch (record.level) {
      case Level.Severe:
        _loggingService.error(...);
    }
});
```

IMO, this has a few drawbacks:

  1. You couple your application to the Logger class directly, instead of hiding the Logger behind a class you control. This becomes a problem if you need add more information to your logs.
  2. You still need to *do* something with those records, and doing something becomes more complex -- generally nested switch or if statements. 
  3. Difficult to trace the execution path. With `loggingService.info` you can jump-to-definition to see what it's doing. You cannot do the same with `Logger().info` -- you need to find the correct `onRecord.listen` and trace through the nested ifs/switches.

Therefore, I've always found it easier to leave the logging package out of the apps I work on and work directly with my own Repository or Service. What do you think?

brianegan replied:
@felangel Actually I'd be curious to hear how you all handle logging. Do you make use of the Logging package? If so, how do you structure it?

felangel replied:
@brianegan it's funny you should ask. In the past we've made use of the [lumberdash](https://pub.dev/packages/lumberdash) logging client but for various reasons we've decided to move away from it and are in the process of evaluating alternative approaches 😄 

With lumberdash, we'd initialize different loggers depending on if the app was running in debug or release mode. For example, in release mode we might use [firebase_lumberdash](https://pub.dev/packages/firebase_lumberdash) or [sentry_lumberdash](https://pub.dev/packages/sentry_lumberdash) whereas in debug mode we might just use [print_lumberdash](https://pub.dev/packages/print_lumberdash).

fwiw I've been playing around with using the [built-in log api](https://api.flutter.dev/flutter/dart-developer/log.html) to see if that could be extended to support these various use-cases.

brianegan replied:
Ah, Lumberdash is a very nice package as well :) Good to hear we're thinking of similar use-cases. 

Thanks again for your help -- really appreciate your time and feedback!

filiph commented on proposals/proposal_1/lib/src/settings/app_theme.dart:
```diff
@@ -0,0 +1 @@
+enum AppTheme { light, dark }
```
If we wanted to go that way, and make `Settings` even closer to production-ready, we could do `{ light, dark, system}` here and set it up. It's something most app will want to do.

felangel replied:
Isn't this equivalent to https://api.flutter.dev/flutter/material/ThemeMode-class.html?

brianegan replied:
It's a fair point -- I wanted this to be "in the control of the developer" in case they wanted to add additional, custom themes as well -- think Slack with lots of theme options.

However, the naming is bad here. I'll look at how to simplify this, or if we should just store the ThemeMode.

brianegan replied:
Switched it up to ThemeMode for now! Thanks again for the suggestion :)

filiph commented on proposals/proposal_1/lib/src/settings/settings_service.dart:
```diff
@@ -0,0 +1,13 @@
+// By default, this class does not persist user settings. If you'd like to
+// persist the user settings locally, use the shared_preferences package to
+// update the methods below. If you'd like to store settings on a web
+// server, use the http package.
```
Excellent!

filiph commented on proposals/proposal_1/pubspec.yaml:
```diff
@@ -0,0 +1,79 @@
+name: proposal_1
+description: A new Flutter project.
+
+# The following line prevents the package from being accidentally published to
```
Not urgent, but this file should definitely get an overhaul. Remove most comments, for example.

brianegan replied:
For sure. I've created a list of "Refinements" that need to happen!

brianegan replied:
I've looked through the pubspec.yaml  and made changes. It's ready for review.

filiph commented on proposals/proposal_1/test/unit_test.dart:
```diff
@@ -0,0 +1,44 @@
+
+void main() {
+  group('MyController', () {
+    testWidgets('should start at 0', (WidgetTester tester) async {
+      // Arrange
+      final myController = MyController();
+
+      // Assert
+      expect(myController.count, 0);
+    });
+
+    testWidgets('should increment by 1', (WidgetTester tester) async {
+      // Arrange
+      final myController = MyController();
+
+      // Act
+      myController.increment();
+
+      // Assert
+      expect(myController.count, 1);
+    });
+  });
```
I think this is convoluted, especially since it's going to get removed. Either test something like `expect(1+1, 2);`, or remove entirely. It's definitely something for a codelab / sample, though.

brianegan replied:
Makes sense -- I'd argue for deleting the file entirely in that case in order to follow the "Can we build the functionality in a way that 90% of developers will want to keep it and/or add to it, instead of ripping it out?" rule.

Sound good?

filiph replied:
I do think that most devs will want to keep the file and add their tests to it. If they have to replace the `expect(1+1, notEqual(3));` line in the process, I think that's fine. WDYT?

filiph reviewed (approved):
Thank you!

I've read through it and added some comment.

I think we're ready to fine-tune what's in the template vs what's in the (inevitable) sample/codelab.

Basically, there is functionality we might still need to rip out, and there is functionality we might still want to expand. I'd say there are some things that _almost every_ app will want to have implemented, and for which there will be few needs for customization (like `AppTheme.system`, see one of the inline comments below).

I'd like to ask you to think about other things like that. The litmus test is:

1) Do most apps need it?
2) Can we build the functionality in a way that 90% of developers will want to keep it and/or add to it, instead of ripping it out?

If the answer is no on any of the above, then it goes out.

And, after seeing the code, I think that:

* Settings is something that is needed in a majority of apps, and you were able to implement it in a way that's simple, and extensible.
* Internationalization, too
* Logging, too
* Also, adaptability

On the other hand, I wonder if we really need `Item` to work. I'd like to see a version of the template where Item/Dummy is only a scaffold, with zero functionality. No plumbing that will need to be ripped out immediately. List view always shows a few items, and detail view always shows the same. That way, the developer can immediately start building _their_ app. But the template gives them a nudge as to how to structure it, and it gives them some important functionality for free.

felangel commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+    // Begin loading the users when the Item List is first shown.
+    widget.controller.initialize();
+
+    super.initState();
```
minor but shouldn't `super.initState()` generally be called first when overriding `initState`?

felangel commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+    );
+  }
+
+  Widget _buildMobileContent(BuildContext context) {
```
shouldn't we avoid using functions and instead extract this to it's own `StatelessWidget`? (https://github.com/dart-lang/language/pull/1450/files/898e538a8d0749dce922de06b92d9bcaf32f315b#r575594530)

felangel commented on proposals/proposal_1/lib/src/items/item_list.dart:
```diff
@@ -0,0 +1,139 @@
+    );
+  }
+
+  Widget _buildTabletContent(BuildContext context) {
```
Shouldn't we avoid using functions to return widgets (https://github.com/dart-lang/language/pull/1450/files/898e538a8d0749dce922de06b92d9bcaf32f315b#r575594530)? I understand why it's being done in this case but it might not be clear why there's a combination of private `StatelessWidgets` like `_ListView` and methods like `_buildTabletContent`. Maybe it would be good to have some inline documentation explaining when to prefer classes as opposed to methods?

brianegan replied:
Yeah for sure! I'm unclear whether we should keep some of this responsive stuff in at all, so this code is definitely WIP :) I've added a "Refinements" section above to make sure I track this one!

brianegan commented:
Thanks for taking an early look and giving thoughts on the overall direction! 

Definitely some things that are more polished, and other parts are more questionable WIP code in here that needs a clean-up :) 

I'll go through all the feedback and make updates on Monday and Tuesday!

brianegan commented:
> Can we build the functionality in a way that 90% of developers will want to keep it and/or add to it, instead of ripping it out?

  > On the other hand, I wonder if we really need Item to work. I'd like to see a version of the template where Item/Dummy is only a scaffold, with zero functionality. No plumbing that will need to be ripped out immediately.

This was my gut feeling as well -- the `items` folder contains lot of code, most of which will be deleted. In my opinion, even handling responsive breakpoints really comes down to the design of the app and it's very hard to put anything general in place.

I'll try a rendition with a much simpler `items` or `dummy` folder to see how it goes, keeping in mind the "90%" rule.

brianegan committed:
Update proposals/proposal_1/analysis_options.yaml

Co-authored-by: Filip Hracek <filiph@users.noreply.github.com>

brianegan committed:
Update proposals/proposal_1/lib/src/items/item_controller.dart

Co-authored-by: Filip Hracek <filiph@users.noreply.github.com>

brianegan committed:
Update proposals/proposal_1/lib/src/items/item_list.dart

Co-authored-by: Filip Hracek <filiph@users.noreply.github.com>

brianegan committed:
Update proposal #1 based on feedback.

brianegan committed:
Update based on feedback

  - Remove Logging
  - Remove Breakpoints
  - Rename Item -> DummyItem
  - Rename item folder -> dummy_feature
  - Remove ItemController
  - Add code comments throughout codebase
  - Clean up pubspec.yaml
  - Remove build methods, only use Widgets
  - Finalize analysis options

brianegan commented:
Thanks again for all the feedback! I've published another version of the proposal that is starting to get into a bit more production-ready shape. I'd appreciate any feedback folks have the time to give!

Changes include:

  - Remove Logging. Different use cases may warrant different samples or documentation.
  - Remove Responsive Breakpoints. Design highly likely to change. Too much code is hinderence.
  - Rename Item -> DummyItem
  - Rename item folder -> dummy_feature
  - Remove ItemController
  - Add code comments throughout codebase
  - Clean up pubspec.yaml
  - Remove build methods, only use Widgets
  - Finalize analysis options

brianegan committed:
Add periods.

brianegan committed:
Add moar periods.

filiph commented:
When I checked out the PR just now and ran `flutter pub get`, I get an error:

```
/Users/filiph/fvm/versions/stable/bin/flutter --no-color pub get
Resource attribute "@appTitle" was not found. Please ensure that each resource has a corresponding @resource.
Generating synthetic localizations package has failed.
Process finished with exit code 1
```

Something with i18n? Just thought I'd mention it.

brianegan commented:
Ah, interesting -- I haven't run into that error myself, but I've added a `@appTitle` key/value pair to the arb file just in case.

I'd guess we're running slightly different versions of Flutter :)

filiph commented:
I still can't get this to work on `stable`. The i18n files won't generate. Switching to `dev` to see if that helps.

brianegan committed:
Fix translations for stable

brianegan commented:
Sorry about that, Filip! Totally my bad -- I'd committed the change and thought I'd pushed it before our meeting, but I mistakenly hadn't. Hopefully work now!

brianegan committed:
Add README for assets & localizations

brianegan commented:
Thanks again for the feedback, everyone! I'll merge this up and close it out. I've created an official proposal on the Flutter repo if you'd like to follow the conversation further :)

https://github.com/flutter/flutter/pull/76519
